include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -1171,13 +1442,31 @@ namespace VectorTools
                 const Quadrature<dim-1>        &q_boundary,
                 const bool                     project_to_boundary_first)
   {
-    do_project (mapping, dof, constraints, quadrature,
-                function, vec_result,
-                enforce_zero_boundary, q_boundary,
-                project_to_boundary_first);
+    if (dim==spacedim)
+      {
+        const Mapping<dim> *const mapping_ptr = dynamic_cast<const Mapping<dim>*> (&mapping);
```

Comment:
The problem is that we can't handle `dim!=spacedim` with the new implementation, but we instantiate also this case. The compiler complains that `Mapping<dim,spacedim>` can't be used in the matrix-free code despite the switch.